components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java

```diff
@@ -236,4 +243,77 @@ public String getRequestOrigin() {
+            throws APISecurityException {
+
+        Map headers =
+                (Map) axis2MessageContext.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+        Object sslCertObject = axis2MessageContext.getProperty(NhttpConstants.SSL_CLIENT_AUTH_CERT_X509);
+        X509Certificate certificateFromMessageContext = null;
+        X509Certificate certificateFromHeader = null;
+        if (sslCertObject != null) {
+            X509Certificate[] certs = (X509Certificate[]) sslCertObject;
+            certificateFromMessageContext = certs[0];
+        }
+        if (headers.containsKey(Utils.getClientCertificateHeader())) {
+            String base64EncodedCertificate =
+                    (String) headers.get(Utils.getClientCertificateHeader());
+            if (base64EncodedCertificate != null) {
+                base64EncodedCertificate = URLDecoder.decode(base64EncodedCertificate).
+                        replaceAll(APIMgtGatewayConstants.BEGIN_CERTIFICATE_STRING, "")
+                        .replaceAll(APIMgtGatewayConstants.END_CERTIFICATE_STRING, "");
+
+                byte[] bytes = Base64.decodeBase64(base64EncodedCertificate);
+                try (InputStream inputStream = new ByteArrayInputStream(bytes)) {
+                    X509Certificate x509Certificate = X509Certificate.getInstance(inputStream);
+                    if (APIUtil.isCertificateExistInTrustStore(x509Certificate)) {
+                        certificateFromHeader = x509Certificate;
+                    }
+                } catch (IOException | CertificateException | APIManagementException e) {
+                    log.error("Error while Convert into X509Certificate", e);
+                    throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                            "Error while Convert into X509Certificate", e);
+                }
+            }
+        }
+        // Certificate available in header
+        if (certificateFromHeader != null) {
+            if (isClientCertificateValidationDisable()){
+                log.debug("Client Certificate Validation disable hence returning client certificate from header");
+                return certificateFromHeader;
+            }
+            try {
+                if (!APIUtil.isCertificateExistInTrustStore(certificateFromMessageContext)) {
+                    log.error("Certificate in messageContext not available in truststore");
+                    return null;
+                } else {
+                    return certificateFromHeader;
+                }
+            } catch (APIManagementException e) {
+                throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                        "Error while Convert into X509Certificate", e);
+            }
+        } else {
+            return certificateFromMessageContext;
+        }
+    }
+
+    private boolean isClientCertificateValidationDisable() {
```

Comment:
shall we change this to enable to match with `apimgt.mutual_ssl.enable_client_validation`?